File: android/src/main/java/com/lightningkite/khrysalis/location/Geocoding.actual.kt
package com.lightningkite.khrysalis.location

import android.location.Geocoder
import com.lightningkite.khrysalis.post
import com.lightningkite.khrysalis.views.DialogRequest
import com.lightningkite.khrysalis.views.ViewDependency
import com.lightningkite.khrysalis.views.ViewStringRaw
import com.lightningkite.khrysalis.views.showDialog
import io.reactivex.Single

@Deprecated("Use the new RX Style instead")
fun ViewDependency.geocode(
    coordinate: GeoCoordinate,
    onResult: (List<GeoAddress>) -> Unit
) {
    if (coordinate.latitude == 0.0 && coordinate.longitude == 0.0) {
        onResult(listOf())
        return
    }
    Thread {
        try {
            val result = Geocoder(context)
                .getFromLocation(coordinate.latitude, coordinate.longitude, 1)
            post {
                onResult(result.map { it -> it.toKhrysalis() })
            }
        } catch (e: Exception) {
            showDialog(DialogRequest(string = ViewStringRaw(e.message ?: "An unknown error occured")))
        }
    }.start()
}

@Deprecated("Use the new RX Style instead")
fun ViewDependency.geocode(
    address: String,
    onResult: (List<GeoAddress>) -> Unit
) {
    if (address.isEmpty()) {
        onResult(listOf())
        return
    }
    Thread {
        try {
            val result = Geocoder(context)
                .getFromLocationName(address, 1)
            post {
                onResult(result.map { it -> it.toKhrysalis() })
            }
        } catch (e: Exception) {
            showDialog(ViewStringRaw(e.message ?: "An unknown error occured"))
        }
    }.start()
}

fun ViewDependency.geocode(
    address: String,
    maxResults: Int = 1
): Single<List<GeoAddress>> {
    if (address.isEmpty()) {
        return Single.just(listOf())
    }
    return Single.create { emitter ->
        Thread {
            try {
                emitter.onSuccess(Geocoder(context)
                    .getFromLocationName(address, maxResults).map { it -> it.toKhrysalis() })
            } catch (e: Exception) {
                emitter.onError(e)
            }
        }.start()
    }
}

fun ViewDependency.geocode(
    coordinate: GeoCoordinate,
    maxResults: Int = 1
): Single<List<GeoAddress>> {
    if (coordinate.latitude == 0.0 && coordinate.longitude == 0.0) {
        return Single.just(listOf())
    }
    return Single.create { emitter ->
        Thread {
            try {
                emitter.onSuccess(Geocoder(context)
                    .getFromLocation(coordinate.latitude, coordinate.longitude, maxResults).map { it -> it.toKhrysalis() })

            } catch (e: Exception) {
                emitter.onError(e)
            }
        }.start()
    }
}

File: ios/Khrysalis/src/location/Geocoding.actual.swift
//Stub file made with Khrysalis 2 (by Lightning Kite)
import Foundation
import CoreLocation
import RxSwift

public extension ViewDependency {

    //--- ViewDependency.geocode(GeoCoordinate, (List<GeoAddress>)->Unit)
    func geocode(_ latLng: GeoCoordinate, _ onResult: @escaping (Array<GeoAddress>)->Void) {
        geocode(latLng: latLng, onResult: onResult)
    }
    func geocode(latLng: GeoCoordinate, onResult: @escaping (Array<GeoAddress>)->Void) {
        CLGeocoder().reverseGeocodeLocation(CLLocation(latitude: latLng.latitude, longitude: latLng.longitude)){ marks, error in
            onResult(marks?.map { translate(mark: $0) } ?? [])
        }
    }

    //--- ViewDependency.geocode(String, (List<GeoAddress>)->Unit)
    func geocode(_ address: String, _ onResult: @escaping (Array<GeoAddress>)->Void) {
        geocode(address: address, onResult: onResult)
    }
    func geocode(address: String, onResult: @escaping (Array<GeoAddress>)->Void) {
        CLGeocoder().geocodeAddressString(address){ marks, error in
            onResult(marks?.map { translate(mark: $0) } ?? [])
        }
    }
    
    func geocode(_ address:String, _ maxResults:Int = 1) -> Single<Array<GeoAddress>> {
        return geocode(address:address, maxResults: maxResults)
    }
    
    func geocode(address:String, maxResults:Int = 1) -> Single<Array<GeoAddress>> {
        if address.isEmpty {
            return Single.just(Array())
        }
        return Single.create ({(emitter: SingleEmitter<Array<GeoAddress>>)in
            CLGeocoder().geocodeAddressString(address){ marks, error in
                emitter.onSuccess(marks?.map { translate(mark: $0) } ?? [])
            }
        })
    }
    
    func geocode(_ coordinate: GeoCoordinate) -> Single<Array<GeoAddress>> {
        return geocode(coordinate:coordinate)
    }
    
    func geocode(coordinate: GeoCoordinate) -> Single<Array<GeoAddress>> {
        return Single.create ({(emitter: SingleEmitter<Array<GeoAddress>>)in
            CLGeocoder().reverseGeocodeLocation(CLLocation(latitude: coordinate.latitude, longitude: coordinate.longitude)){ marks, error in
                emitter.onSuccess(marks?.map { translate(mark: $0) } ?? [])
            }
        })
    }
}

//--- translate
private func translate(mark: CLPlacemark) -> GeoAddress {
    return GeoAddress(coordinate: mark.location?.coordinate.toKhrysalis(), name: mark.name, street: mark.street, subLocality: mark.subLocality, locality: mark.locality, subAdminArea: mark.subAdministrativeArea, adminArea: mark.administrativeArea, countryName: mark.country, postalCode: mark.postalCode)
}

private extension CLPlacemark {
    
    var street: String? {
        if let x = subThoroughfare, let y = thoroughfare {
            return "\(x) \(y)"
        }
        return nil
    }
}
